File: pc02/pc02_07/python/pc02_07.py
"""
Contests scoreboard
pc110207
uva10258
"""

import sys


class Contestant:
    """Represents a contest contestant."""

    def __init__(self):
        self.solved = 0
        self.penalty = 0
        self.participated = False
        self.wrong = [0] * 10
        self.solvedFlag = [False] * 10

    def submit(self, problem, time, verdict):
        """Process a submission."""
        self.participated = True

        if verdict == 'I':
            if not self.solvedFlag[problem]:
                self.wrong[problem] += 1
        elif verdict == 'C':
            if not self.solvedFlag[problem]:
                self.solvedFlag[problem] = True
                self.solved += 1
                self.penalty += time + 20 * self.wrong[problem]


def main():
    T = int(sys.stdin.readline())

    first = True

    for _ in range(T):
        if not first:
            print()
        first = False

        teams = [Contestant() for _ in range(101)]

        for line in sys.stdin:
            line = line.strip()
            if not line:
                break

            parts = line.split()
            t = int(parts[0])
            p = int(parts[1])
            time = int(parts[2])
            verdict = parts[3]

            teams[t].submit(p, time, verdict)

        scoreboard = []
        for i in range(1, 101):
            if teams[i].participated:
                scoreboard.append((i, teams[i].solved, teams[i].penalty))

        scoreboard.sort(key=lambda x: (-x[1], x[2], x[0]))

        for team_id, solved, penalty in scoreboard:
            print(f"{team_id} {solved} {penalty}")


if __name__ == '__main__':
    main()

File: pc02/pc02_07/c/pc02_07.c
/**
 * Contests scoreboard
 * pc110207
 * uva10258
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

typedef struct {
	int id;
	int solved;
	int penalty;
} contestant_result_t;

typedef struct {
	int solved;
	int penalty;
	int participated;
	int wrong[10];
	int solvedFlag[10];
} contestant_t;

int compare_contestants(const void *a, const void *b) {
	contestant_result_t *ca = (contestant_result_t *)a;
	contestant_result_t *cb = (contestant_result_t *)b;

	if (cb->solved != ca->solved) {
		return cb->solved - ca->solved;
	}

	if (ca->penalty != cb->penalty) {
		return ca->penalty - cb->penalty;
	}

	return ca->id - cb->id;
}

void contestant_init(contestant_t *c) {
	c->solved = 0;
	c->penalty = 0;
	c->participated = 0;
	memset(c->wrong, 0, sizeof(c->wrong));
	memset(c->solvedFlag, 0, sizeof(c->solvedFlag));
}

void contestant_submit(contestant_t *c, int problem, int time, char verdict) {
	c->participated = 1;
	int idx = problem - 1;

	if (verdict == 'I') {
		if (!c->solvedFlag[idx]) {
			c->wrong[idx]++;
		}
	} else if (verdict == 'C') {
		if (!c->solvedFlag[idx]) {
			c->solvedFlag[idx] = 1;
			c->solved++;
			c->penalty += time + 20 * c->wrong[idx];
		}
	}
}

int main(void) {
	int T;
	scanf("%d", &T);

	int is_first = 1;

	while (T--) {
		if (!is_first) {
			printf("\n");
		}
		is_first = 0;

		contestant_t teams[101];
		for (int i = 0; i <= 100; i++) {
			contestant_init(&teams[i]);
		}

		char line[256];
		while (fgets(line, sizeof(line), stdin)) {
			if (line[0] == '\n') {
				break;
			}

			int t, p, time;
			char verdict;

			if (sscanf(line, "%d%d%d %c", &t, &p, &time, &verdict) == 4) {
				contestant_submit(&teams[t], p, time, verdict);
			}
		}

		contestant_result_t scoreboard[101];
		int count = 0;

		for (int i = 1; i <= 100; i++) {
			if (teams[i].participated) {
				scoreboard[count].id = i;
				scoreboard[count].solved = teams[i].solved;
				scoreboard[count].penalty = teams[i].penalty;
				count++;
			}
		}

		qsort(scoreboard, count, sizeof(contestant_result_t), compare_contestants);

		for (int i = 0; i < count; i++) {
			printf("%d %d %d\n", scoreboard[i].id, scoreboard[i].solved, scoreboard[i].penalty);
		}
	}

	return 0;
}
